### PR TITLE
Use AutoDiff for quaternion functions, new gradient utility functions

### PR DESCRIPTION
**Pull request by tkoolen**
Sorry for doing more than one thing in a pull request. This started off as just switching the quaternion stuff over to AutoDiff (which I later need in order to implement a new type of orientation trajectory that Sylvain from IHMC showed me), but I realized I kind of needed some new utility functions in the process.

### CODE REVIEW

File: drake/core/Gradient.h [hunk 1]
```diff
@@ -18,55 +18,158 @@ namespace Drake {
...
+    initializeAutoDiff(mat, ret, num_derivatives, deriv_num_start);
+    return ret;
   }
 
-  template <typename Derived>
-  TaylorVecXd constantTaylorVecXd(const Eigen::MatrixBase<Derived>& val) {
-    TaylorVecXd auto_diff_vec(val.rows());
-    constantTaylorVecXd(val, auto_diff_vec);
-    return auto_diff_vec;
+  // TODO begin: move to different location?
```
Comment (tkoolen) [inline drake/core/Gradient.h]: Yeah, so is there a better place for this. At first I thought Vector.h, but I need the SizeAtCompileTime, not the RowsAtCompileTime. We could switch the Vector stuff over to using SizeAtCompileTime.

File: drake/solvers/Optimization.h [hunk 2]
```diff
@@ -130,7 +130,7 @@ namespace Drake {
     virtual ~QuadraticConstraint() {}
 
     virtual void eval(const Eigen::Ref<const Eigen::VectorXd>& x, Eigen::VectorXd& y) const override { y.resize(getNumConstraints()); y=.5*x.transpose()*Q*x + b.transpose()*x; }
-    virtual void eval(const Eigen::Ref<const TaylorVecXd>& x, TaylorVecXd& y) const override { y.resize(getNumConstraints()); y = .5*x.transpose()*constantTaylorVecXd(Q)*x + constantTaylorVecXd(b).transpose()*x; };
+    virtual void eval(const Eigen::Ref<const TaylorVecXd>& x, TaylorVecXd& y) const override { y.resize(getNumConstraints()); y = .5*x.transpose()* Q.cast<TaylorVarXd>() * x + b.cast<TaylorVarXd>().transpose() * x; };
```
Comment (tkoolen) [inline drake/solvers/Optimization.h]: Could undo this change if you want, but I thought it was a bit unnecessary (after having cleared up my confusion).

File: drake/systems/controllers/LQR.h [hunk 3]
```diff
@@ -17,15 +17,11 @@ namespace Drake {
...
-    // todo: clean this up
-    typedef TaylorVard<-1> AutoDiffType;
-    VectorXd xu(num_states+num_inputs);
-    xu << toEigen(x0), toEigen(u0);
-    TaylorVecXd xu_taylor = initTaylorVecXd(xu);
-    typename System::template StateVector<AutoDiffType> x_taylor(xu_taylor.head(num_states));
-    typename System::template InputVector<AutoDiffType> u_taylor(xu_taylor.tail(num_inputs));
-
-    auto xdot = autoDiffToGradientMatrix(toEigen(sys.dynamics(AutoDiffType(0),x_taylor,u_taylor)));
+    auto autodiff_args = initializeAutoDiffArgs(toEigen(x0), toEigen(u0));
```
Comment (tkoolen) [inline drake/systems/controllers/LQR.h]: So this is what the new stuff in Gradient.h allows you to do. Given a bunch of Eigen matrix arguments, construct a tuple of corresponding AutoDiff matrices, with values and derivatives properly initialized.

File: drake/util/test/testQuatmex.cpp [hunk 4]
```diff
@@ -1,52 +1,73 @@
...
 
-  plhs[4] = mxCreateDoubleMatrix(4,1,mxREAL);
-  plhs[5] = mxCreateDoubleMatrix(4,8,mxREAL);
-  memcpy(mxGetPr(plhs[4]),q3.data(),sizeof(double)*4);
-  memcpy(mxGetPr(plhs[5]),dq3.data(),sizeof(double)*4*8);
+  {
+    auto autodiff_args = initializeAutoDiffArgs(q1, q2);
+    auto r_autodiff = quatDiff(get<0>(autodiff_args), get<1>(autodiff_args));
+    auto r = autoDiffToValueMatrix(r_autodiff);
+    auto dr = autoDiffToGradientMatrix(r_autodiff);
```
Comment (tkoolen) [inline drake/util/test/testQuatmex.cpp]: Another example. I think it's quite nice.

**Comment by tkoolen** [comment]
compiles (with -Wenum warnings) on g++ 4.9.3.

File: drake/core/Gradient.h [hunk 5]
```diff
@@ -18,55 +18,153 @@ namespace Drake {
...
+      auto_diff_matrix(row).derivatives() = gradient.row(row).transpose();
     }
   }
 
-  template <typename Derived>
-  TaylorVecXd initTaylorVecXd(const Eigen::MatrixBase<Derived>& val) {
-    TaylorVecXd auto_diff_vec(val.rows());
-    initTaylorVecXd(val, auto_diff_vec);
-    return auto_diff_vec;
+  /** \brief initialize a single autodiff matrix given the corresponding value matrix.
```
Comment (RussTedrake) [inline drake/core/Gradient.h]: please document very clearly here what happens to the gradient vector.  eye()?  zero()?

File: drake/core/Gradient.h [hunk 6]
```diff
@@ -18,55 +18,153 @@ namespace Drake {
...
+    auto_diff_matrix.resize(val.rows(), val.cols());
+    Eigen::DenseIndex deriv_num = deriv_num_start;
+    for (DenseIndex i = 0; i < val.size(); i++) {
+      auto_diff_matrix(i) = ADScalar(val(i), num_derivatives, deriv_num++);
+    }
   }
 
-  /** constantTaylorVecXd
-   * @brief initializes the vector with x=val and dx=zeros(numel(val))
+  /** \brief initialize a single autodiff matrix given the corresponding value matrix.
```
Comment (RussTedrake) [inline drake/core/Gradient.h]: document gradient vector.

File: drake/core/Gradient.h [hunk 7]
```diff
@@ -18,55 +18,153 @@ namespace Drake {
...
-    for (Index row = 0; row < auto_diff_vector.size(); row++) {
-      auto_diff_vector(row).value() = val(row);
-      auto_diff_vector(row).derivatives().resize(nx,1);
-      auto_diff_vector(row).derivatives() = gradient.row(row).transpose();
-    }
+  template<typename Head, typename ...Tail>
+  Eigen::DenseIndex totalSizeAtRunTime(const Eigen::MatrixBase<Head> &head, const Tail &... tail) {
+    return head.size() + totalSizeAtRunTime(tail...);
+  }
+  // TODO end: move to different location?
```
Comment (RussTedrake) [inline drake/core/Gradient.h]: I do think core/ is the right place for these things.  It seems like there are a lot of small utilities like this that we might write around eigen types.  perhaps we need a header file just for EigenUtils.h or something?

File: drake/core/Gradient.h [hunk 8]
```diff
@@ -18,55 +18,153 @@ namespace Drake {
...
+    template<>
+    struct InitializeAutoDiffArgsHelper<0> {
+      template<typename ...ValueTypes, typename ...AutoDiffTypes>
+      static void run(const std::tuple<ValueTypes...> &values, const std::tuple<AutoDiffTypes...> &auto_diffs, Eigen::DenseIndex num_derivatives, Eigen::DenseIndex deriv_num_start) {
+        // empty
+      }
+    };
+  }
+
+  template<typename ...Args>
```
Comment (RussTedrake) [inline drake/core/Gradient.h]: absolutely imperative that you document this one.  i have absolutely no idea what it's doing from just looking quickly at the code.

File: drake/solvers/Optimization.h (hunk 2 above)
Comment (RussTedrake) [inline drake/solvers/Optimization.h]: that's fine.
i actually think we should implement (in EigenUtils.h, or wherever) our own support for multiplying double matrices with autodiff matrices (from #1312).  it's ridiculuously inefficient to have to allocate a new autodiff matrix, looping through an zeroing all of the gradients, every time we want to do this.

File: drake/solvers/SNOPTWrapper.cpp [hunk 9]
```diff
@@ -51,7 +51,7 @@ static int snopt_userfun(snopt::integer *Status, snopt::integer *n, snopt::doubl
   memset(G,0,sizeof(*n)*sizeof(snopt::doublereal));
 
   // evaluate objective
-  auto tx = initTaylorVecXd(xvec);
+  auto tx = initializeAutoDiff(xvec);
```
Comment (RussTedrake) [inline drake/solvers/SNOPTWrapper.cpp]: We need to decide and agree on what to call all of these things.  My feeling is that having matrix types defined for AutoDiff helps a lot - i had been calling them TaylorVec, TaylorMat, ... I'm ok changing them, but I'm not ok with you writing everything one way and me another.  I think the function signatures should match the types.

File: drake/systems/controllers/LQR.h (hunk 3 above)
Comment (RussTedrake) [inline drake/systems/controllers/LQR.h]: that's very nice.  but still hard to read even here (the previous code was, too).  maybe there is a better name for that method?  even initializeAutoDiffTuple might be better?

File: drake/systems/plants/RigidBodySystem.cpp [hunk 10]
```diff
@@ -185,7 +185,7 @@ class SingleTimeKinematicConstraintWrapper : public Constraint
     rigid_body_constraint->getRobotPointer()->doKinematics(kinsol);
     VectorXd y; MatrixXd dy;
     rigid_body_constraint->eval(nullptr,kinsol,y,dy);
-    initTaylorVecXd(y,dy*autoDiffToGradientMatrix(tq),ty);
+    initializeAutoDiffGivenGradientMatrix(y, (dy * autoDiffToGradientMatrix(tq)).eval(), ty);
```
Comment (RussTedrake) [inline drake/systems/plants/RigidBodySystem.cpp]: by the way, i don't really like that i have to do this everywhere (bring the chain rule inside of my method).  i thought a little about method signatures which took a double in and outputted an autodiff (so the chain rule would be on the outside), but ultimately decided this was cleaner; I think.  When it's templated, everything will work correctly (except that people have to write their templated methods using that ugly cast to scalartype for data matrices).  But I worry that wrapper methods like this will incur bugs where people forget the chain rule component.

Comment (RussTedrake) [inline drake/systems/plants/RigidBodySystem.cpp]: just asking here to make sure we've discussed it once / get a second opinion.  Not requesting any code change.

File: drake/util/drakeGeometryUtil.cpp [hunk 11]
```diff
@@ -19,168 +19,6 @@ double angleDiff(double phi1, double phi2)
   return d;
 }
 
-
-Vector4d quatConjugate(const Eigen::Vector4d& q)
```
Comment (RussTedrake) [inline drake/util/drakeGeometryUtil.cpp]: while it's nice to see all of this go, it does mean more in the headers, etc.  did you evaluate performance?
one of the big things about all of these geometry utils is that they work on fixed-size matrices (and can potentially avoid most dynamic allocation).  Is the same true for your intializeAutodDiffArgs version?  (I think so, but again it's hard to tell... if yes, please include that in the method documentation)

File: drake/util/test/testQuatmex.cpp (hunk 4 above)
Comment (RussTedrake) [inline drake/util/test/testQuatmex.cpp]: that is very nice!

**Comment by RussTedrake** [comment]
finished my review.  lots to discuss, but very very cool.

File: drake/core/Gradient.h (hunk 8 above)
Comment (tkoolen) [inline drake/core/Gradient.h]: done.

File: drake/solvers/SNOPTWrapper.cpp (hunk 9 above)
Comment (tkoolen) [inline drake/solvers/SNOPTWrapper.cpp]: Yeah, maybe we can talk about it in person today?

File: drake/util/drakeGeometryUtil.cpp (hunk 11 above)
Comment (tkoolen) [inline drake/util/drakeGeometryUtil.cpp]: > while it's nice to see all of this go, it does mean more in the headers, etc. 

Almost everything else in drakeGeometryUtil is already in the header, and I think that's where it should be for this low level stuff.

> did you evaluate performance?

No. But it should be exactly the same.

> one of the big things about all of these geometry utils is that they work on fixed-size matrices (and can potentially avoid most dynamic allocation).

Exactly.

> Is the same true for your intializeAutodDiffArgs version?

Yeah, if all of the inputs have fixed size, then the derivative vector will also have fixed size (being the sum of the sizes at compile time of all of the input arguments). Adding this to the documentation. I'd like to handle the fixed max size case too, but later.

**Comment by tkoolen** [comment]
Closed and reopened to re-trigger build servers after resolution of https://github.com/RobotLocomotion/drake/issues/1652. Should be good after that I think (didn't have failing unit tests anymore before).

**Comment by tkoolen** [comment]
I can't seem to reproduce the appveyor build error on drake003 or my own machine. Any ideas?

One of the errors:
```
c:\projects\drake\build\include\eigen3\eigen\src/Core/util/Meta.h(133): error C2719: 'unnamed-parameter': formal parameter with requested alignment of 16 won't be aligned [C:\projects\drake\drake\pod-build\systems\plants\constraint\drakeRigidBodyConstraint.vcxproj] [C:\projects\drake\pod-build\drake.vcxproj]
    c:\projects\drake\build\include\eigen3\eigen\src/Core/CwiseUnaryOp.h(43): note: see reference to class template instantiation 'Eigen::internal::result_of<UnaryOp (Eigen::AutoDiffScalar<Eigen::Matrix<double,8,1,0,8,1>>)>' being compiled
            with
            [
                UnaryOp=Eigen::internal::scalar_multiple_op<Eigen::AutoDiffScalar<Eigen::Matrix<double,8,1,0,8,1>>>
            ]
    c:\projects\drake\build\include\eigen3\eigen\src/Core/CwiseUnaryOp.h(93): note: see reference to class template instantiation 'Eigen::internal::traits<Derived>' being compiled
            with
            [
                Derived=Eigen::CwiseUnaryOp<Eigen::internal::scalar_multiple_op<Eigen::AutoDiffScalar<Eigen::Matrix<double,8,1,0,8,1>>>,const Eigen::Block<const Eigen::Matrix<Eigen::AutoDiffScalar<Eigen::Matrix<double,8,1,0,8,1>>,4,1,0,4,1>,3,1,false>>
            ]
    c:\projects\drake\build\include\eigen3\eigen\src/Core/CwiseUnaryOp.h(61): note: see reference to class template instantiation 'Eigen::CwiseUnaryOpImpl<UnaryOp,XprType,Eigen::Dense>' being compiled
            with
            [
                UnaryOp=Eigen::internal::scalar_multiple_op<Eigen::AutoDiffScalar<Eigen::Matrix<double,8,1,0,8,1>>>,
                XprType=Eigen::Block<const Eigen::Matrix<Eigen::AutoDiffScalar<Eigen::Matrix<double,8,1,0,8,1>>,4,1,0,4,1>,3,1,false>
            ]
    C:\projects\drake\drake\..\drake/util/drakeGeometryUtil.h(52): note: see reference to class template instantiation 'Eigen::CwiseUnaryOp<Eigen::internal::scalar_multiple_op<Eigen::AutoDiffScalar<Eigen::Matrix<double,8,1,0,8,1>>>,const Derived>' being compiled
            with
            [
                Derived=Eigen::Block<const Eigen::Matrix<Eigen::AutoDiffScalar<Eigen::Matrix<double,8,1,0,8,1>>,4,1,0,4,1>,3,1,false>

```

**Comment by RussTedrake** [comment]
That’s the 32-bit build — and it’s a byte alignment issue.  drake004 would error, too

**Comment by tkoolen** [comment]
Sorry, I meant drake004, not drake003. At least I think I did; now I'm starting to doubt whether I was using the right machine for my test. I actually can reproduce the error on my win64 machine in 32 bit cross compiler mode now though.

By the way, what's going on with drake004? I can ssh in from the lab, but then I can't ping github (or google) from drake004.

**Comment by tkoolen** [comment]
For reference, I think this is another case of http://eigen.tuxfamily.org/bz/show_bug.cgi?id=1006, which also occurred in https://github.com/RobotLocomotion/drake/pull/1584. It's because we're using matrices where the scalar type itself is aligned (as it is an AutoDiffScalar with a fixed-size vectorizable derivatives vector). In #1584 I got around it by just using a dynamic size derivatives vector, but I guess I'll need to solve it for real this time.

**Comment by tkoolen** [comment]
Alright, so the issue is in quatProduct. I can get around it by first switching to bleeding edge Eigen (default branch), using Eigen's Quaternion operator* to do the computation, and changing a value to a const reference in Eigen's Translation::isApprox function (for which I've opened a bug report with Eigen, http://eigen.tuxfamily.org/bz/show_bug.cgi?id=1156).

With our own implementation, the scalar_multiple_op created by multiplying a vector by a number in quatProduct creates an issue with Eigen's internal result_of function. I commented on http://eigen.tuxfamily.org/bz/show_bug.cgi?id=96 to report this.

**Comment by tkoolen** [comment]
Alright, I've got almost everything resolved after updating to eigen 3.3 beta; there's just one annoying issue with swig left. @rdeits, I found out what's going wrong, not sure what's the right way to fix it yet though. Once that's resolved, I'm going to first do a pull request to only update eigen to 3.3 beta, making the necessary changes, then I'll rebase this onto the eigen 3.3 beta branch.

**Comment by tkoolen** [comment]
Gael from Eigen has been super helpful by the way, he's fixed both of the issues from my comment above already.

**Comment by tkoolen** [comment]
appveyor win64: failed to download pkg-config.exe.
drake005: I guess Matlab was killed while executing examples/Cartpole/runHinf for some reason.

**Comment by tkoolen** [comment]
test
